Mathlib/RingTheory/Bialgebra/Hom.lean

```diff
@@ -0,0 +1,308 @@
+/-
+Copyright (c) 2024 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau, Yury Kudryashov, Amelia Livingston
+-/
+import Mathlib.RingTheory.Coalgebra.Hom
+import Mathlib.RingTheory.Bialgebra.Basic
+
+/-!
+# Homomorphisms of `R`-bialgebras
+
+This file defines bundled homomorphisms of `R`-bialgebras. We simply mimic
+`Mathlib/Algebra/Algebra/Hom.lean`.
+
+## Main definitions
+
+* `BialgHom R A B`: the type of `R`-bialgebra morphisms from `A` to `B`.
+* `Bialgebra.counitBialgHom R A : A →ₐc[R] R`: the counit of a bialgebra as a bialgebra
+homomorphism.
+
+## Notations
+
+* `A →ₐc[R] B` : `R`-bialgebra homomorphism from `A` to `B`.
+
+-/
+
+open TensorProduct Bialgebra BigOperators
+
+universe u v w
+
+/-- Given `R`-algebras `A, B` with comultiplication maps `Δ_A, Δ_B` and counit maps
+`ε_A, ε_B`, an `R`-bialgebra homomorphism `A →ₐc[R] B` is an `R`-algebra map `f` such that
+`ε_B ∘ f = ε_A` and `(f ⊗ f) ∘ Δ_A = Δ_B ∘ f`. -/
+structure BialgHom (R A B : Type*) [CommSemiring R]
+    [Semiring A] [Algebra R A] [Semiring B] [Algebra R B]
+    [CoalgebraStruct R A] [CoalgebraStruct R B] extends A →ₗc[R] B, A →* B
+
+/-- Reinterpret a `BialgHom` as a `MonoidHom` -/
+add_decl_doc BialgHom.toMonoidHom
+
+@[inherit_doc BialgHom]
+infixr:25 " →ₐc " => BialgHom _
+
+@[inherit_doc]
+notation:25 A " →ₐc[" R "] " B => BialgHom R A B
+
+/-- `BialgHomClass F R A B` asserts `F` is a type of bundled bialgebra homomorphisms
+from `A` to `B`.  -/
+class BialgHomClass (F : Type*) (R A B : outParam Type*)
```

Comment:
Should this file just be exactly following the "best practices" laid out in [this module doc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/FunLike/Basic.html)? Do you know what's going on when you diverge from this template? For example you don't seem to have made `copy` -- should you or doesn't this matter?

Reply:
I've added `CoalgHom.copy` & `BialgHom.copy` now; I don't think anything else mentioned in the link is missing